'accumulate' for Raise

serras opened the pull request:
This PR proposed a one-for-all accumulating map for `Raise`. The signature is not the greatest at this moment, because `Raise` needs to appear in receiver position, so the `list` goes as first argument.

```kotlin
public fun <R, A, B> Raise<NonEmptyList<R>>.accumulate(
  list: Iterable<A>,
  block: Raise<R>.(A) -> B
): List<B>
```

Interestingly, this function is actually generic over the way we run the `Raise`.

raulraja reviewed (approved):
Thanks @serras, looks great !

nomisRev commented:
That signature makes me sad 😞 I'm personally not really in favour of adding this method until we have context receivers. I think the pattern of working over `Raise<E>` will not really catch on they become stable.

You can of course call `bind` here inside of `block`, but then this seems like a micro-optimisation compared to the current signature.

Working `E` doesn't solve any problems, since you typically want `Nel<E>`. The signature below suffers from a second problem which is that it's not `inline`, which is not possible inside interfaces. So it doesn't allow foreign _suspend_ to pass through, which is a must in this model of "transformers". To solve this it would also need to be put outside of the `interface` which makes this suffer from the same issue.

```kotlin
interface Raise<E> {
  fun <A, B> Iterable<A>.mapOrAccumulate(combine: (E,E) -> E, transform: Raise<E>.(A) -> B): List<B>
}
```

nomisRev commented on arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/Raise.kt:
```diff
@@ -247,3 +248,23 @@ public inline fun <R, B : Any> Raise<R>.ensureNotNull(value: B?, raise: () -> R)
+ * over every element of [list].
+ */
+@EffectDSL
+public fun <R, A, B> Raise<NonEmptyList<R>>.accumulate(
```
```suggestion
public inline fun <R, A, B> Raise<NonEmptyList<R>>.accumulate(
```

serras commented:
@nomisRev I agree that context receivers will make this much nicer. However, I don't think that the following looks so terrible,

```kotlin
accumulate(elements) {
  checkValidity(it)
}
```

when compared to

```kotlin
elements.accumulate {
  checkValidity(it)
}
```

nomisRev commented:
Alright, that's fair. Is `accumulate` the correct name for this operator? It still has the shape of `map`, right?

I think we should add the other signature inside `Raise` already since that fits.

```kotlin
interface Raise<in E> {
  fun <A, B> Iterable<A>.accumulate(
    semigroup: Semigroup<E>,
    transform: Raise<E>.(A) -> B
  ): List<B>
}
```

serras commented:
@nomisRev I updated the code to use the more general version you mention; in fact I've added an even more general one because to implement the version with `NonEmptyList<E>` you may want a way to "embed" a single `E` into the non-empty list.

i-walker commented:
Is it possible to instead use a combination of mapLeft and mapAccumulate without the embed parameter. This way the method isnt too convoluted @serras? Unless that is what we were going for

serras commented:
@i-walker indeed, I think you're right. I've pushed a new commit in which `mapOrAccumulate` within `Raise` is simpler, and I use a combination with `mapLeft` to implement the `NonEmptyList` version

nomisRev commented on arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/Raise.kt:
```diff
@@ -62,6 +64,47 @@ public interface Raise<in R> {
+    semigroup: Semigroup<@UnsafeVariance R>,
+    block: Raise<E>.(A) -> B
+  ): List<B> {
+    var error: R? = null
```
Using `null` here is unsafe due to the nested null problem.

serras replied:
I didn't know about `EmptyValue`, this has been fixed now.

i-walker commented on arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/Raise.kt:
```diff
@@ -211,6 +242,16 @@ public interface Raise<in R> {
+ * into one of type [R].
+ */
+@EffectDSL
+public inline fun <R, E, A> Raise<R>.mapLeft(
```
since `mapLeft` is just recover can we remove this? and inline it into `mapOrAccumulate`

serras replied:
Done! I've kept `mapLeft` in the code, because it might be useful elsewhere.

i-walker replied:
The method is still there can we add this in a separate PR :)

serras replied:
Indeed, I've just removed it to make this PR self-contained
